### DISCUSSION TITLE
UMI 4.0 踩坑记录

### DISCUSSION THREAD

**Discussion by huarse** [umijs/umi#8421] [Show and tell]
花了几天时间尝试将原来的 3.0 框架升级到 4.0，确实遇到了不少坑，以下是一些参照官方文档也无法避免的问题，仅供参考：

1. 配置文件中通过 `styles` 和 `scripts` 注入的文件均会在 umi.js 和 umi.css 后面，导致出现依赖顺序问题，解决办法：写一个插件，通过 `modifyHTML` 等 API 手动注入；
2. 配置文件中 `target` 不能有 edge 浏览器，否则会报错；
3. 配置路由时，`redirect` 不能带上路径参数（已经有人提 [PR](https://github.com/umijs/umi/pull/8395)，等待解决，可临时在入口加手动跳转）；
4. `history` 设置为 `browser` ，并开启 `runtimePublicPath` 时，在 development 环境下，`window.publicPath` 必须设置成 `/`，如果是 `./` 会导致依赖文件都从当前的相对路径加载；
5. 路由配置中的 title 不再生效，需要在 app.tsx (入口文件) 中监听 `onRouteChange`，然后手动修改 `document.title`
6. 偶尔会出现资源一直加载失败的情况，删除 src/.umi 都不管用，必须将 node_modules/.cache 目录也删掉
7. 工程中中引入了一些 React 组件依赖了老版本的路由API，会出现冲突，导致页面直接挂掉（**此问题暂时无解**）

**Comment by CODEOrz** [comment]
umi从3.5.20 升级到 4.0.0以后
yarn install后 执行 max setup 时，提示此错误
<img width="809" alt="image" src="https://user-images.githubusercontent.com/29497017/179271415-222c0d13-103c-406c-a81f-1d9a2e9449e1.png">

**Reply by fz6m** [reply]
旧版插件 api 不支持了，部分插件在 `@umijs/max` 里 [自带](https://umijs.org/docs/max/introduce) ，其他的可通过：

```bash
  pnpm i -D @umijs/plugins
```

```ts
// .umirc.ts

export default {
  plugins: [
    '@umijs/plugins/dist/${name}'
  ]
}
```

引入。

**Reply by CODEOrz** [reply]
感谢🙏

**Reply by 763483000** [reply]
runtimePublicPath 我现在无法开启，报对象类型错误，请问你是如何开启的，官网只说了是 Object类型，没告诉如何写

**Reply by fz6m** [reply]
```ts
// .umirc.ts

export default {
  runtimePublicPath: {}
}
```

这个选项主要是用在 qiankun 子应用的，如果用了 qiankun 插件他会自动开启。

**Reply by 763483000** [reply]
不好意思，我不了解qiankun 这个应用 ，我当前之前历史版本的时候都是采用如下方式，我主要是java工程端，我按照如下方式来引入umi.js和umi.css ，但是umi里依赖的js 路径是通过 runtimePublicPath配置的
![image](https://user-images.githubusercontent.com/6050691/179510415-14751134-240b-4b76-8997-2b296a297b66.png)

**Reply by 763483000** [reply]
我已经解决了，我虽然不确定如何解决的，但是我反复尝试
  mfsu: {
    runtimePublicPath: true,
  },
添加了以上配置 ，部署发布，可以了

**Comment by dqm07** [comment]
umi4的时候，需要在html手动配置window.publicPath，然后umirc中配置runtimePublicPath: {}即可
